Sites Management Page: Replace react-intersection-observer with IntersectionObserver

mpkelly opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/68625

#### Proposed Changes

* Remove `react-intersection-observer` and reuse some existing logic elsewhere in the codebase

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

This hook is used on the `/sites` page and at for a nudge at `/activity-log/:business-or-pro-site-with-jetpack`. 

* To test, go to `/sites` with one or more unlaunched sites. Note how many Tracks events fire with `calypso_sites_dashboard_site_launch_nag_inview` in both **list** and **grid** views. Repeat the process on the Calypso Live site and verify the new code produces the same output. The events should only ever fire one and you can verify this by scrolling the nag UI in and out of the screen. 
* Testing `/activity-log/:business-or-pro-site-with-jetpack` is a bit awkward. The easiest way to check is to modify the code to force the banner to display. You can do this like so:
    - go to `client/my-sites/activity/activity-log/index.jsx` and change line `551` so that `<VisibleDaysLimitUpsell/>` always get rendered. 
    - Next, go to `<VisibleDaysLimitUpsell/>` and change line `60` so it doesn't return `null`. Finally, go to `/activity-log/business-or-pro-site-with-jetpack` and scroll to the bottom and should see the banner. As before, ensure the Tracks event only fires one per page load after switching from `trunk` to `remove/react-intersection-observer-dep`

![image](https://user-images.githubusercontent.com/6851384/196109518-a0e9ba7a-3be8-4fad-9195-4b8a291c3e8e.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

mpkelly committed:
Reuse existing code for firing events when Element is fully visible.

mpkelly committed:
Match previous variable name

mpkelly committed:
Fix yarn.lock

danielbachhuber reviewed (commented):
Thanks for working on this, @mpkelly 

Personally, I still think we should stick with the original dependency:
1. It has tests, which this internal implementation does not.
2. Invariably, other people will want to add new features to the implementation — thus, reinventing the wheel.

I'll defer to @p-jackson's decision, though.

zaguiini commented:
Thanks for the contribution, I really appreciate the work on the back end! 🤝

I'm curious to understand what was the decision-making process behind this PR... I don't think it was motivated by bundle size reduction since it removes 2kb from the gzipped distribution. It was not from a maintenance perspective either since the dependency was last updated 3 months ago, plus it's well-tested and supported by the community. 🤔 Happy to approve this once we have better reasoning, but as of right now, it doesn't look like a big deal.

mpkelly commented:
I already laid out my case in the related issue: we already had the code that did what we wanted, so adding an extra dependency didn't seem warranted. I guess this was @p-jackson's thinking too?

@danielbachhuber, @zaguiini, I would say it's more about maintaining a strict dependency policy rather than a one-time saving of 1.8kb. Yes, we only save 1.8kb here, but with a strict policy, we keep the total dependencies and bundle sizes down. Conversely, if we let every small bundle through, we will soon have lots of dependencies to maintain, and more code going through to our users than necessary. Furthermore, the minified code for the hook I added is 600 bytes, so the library is 300% bigger, which is quite a lot of extra code if we consistently use "it's only a small bundle" logic.

p-jackson commented:
I think I'm echoing some of what @mpkelly has already said, but here goes:

> I'm curious to understand what was the decision-making process behind this PR

While I was reviewing #68533 (which is where the dependency was added) I was wary of adding a dependency to achieve the "in view" behaviour, but rather than hold up the review I created the #68625 task, so that we could come back and think about it later. So that's why the ticket is a bit light on details sorry.

**Bundle size**: I agree that 1.8kb is pretty good. My guess is the package does a good job of "tree shaking" so we're only getting what we need. On the other hand, the `useInView` has a lot of options which we don't use. None of that code can be pruned away.
1.8kb is good, but it's all relative. In the context of our megabyte-sized bundle it's good. In the context of a 50kb bundle it'd be pretty bad. Bundle size wasn't really my biggest concern though, except for how it relates to duplication.

**Duplication**: We already have a bunch of code that does intersection observation in Calypso. Shouldn't we have factored out that code into something usable by SMP? Or, if we think the package code is better than our code shouldn't we replace all the existing intersection observers with `useInView`?

**Maintenance**: `react-intersection-observer` does look well maintained, but I'm more worried about our maintenance burden when it comes to dependency updates. We all know updating `package.json` to get a set of versions that all work nicely together can be a nightmare. This would add another thing to the mix with a peer dependency on React. Does it support React 18? Does it work in strict mode? Will there be a breaking API change that we can't put off because it's the only version compatible with all our other dependencies?
The dependency burden isn't only measured in kb, but also in the absolute number of dependencies.

**VanillaJS**
This is more of a principle than a hard reason not to use `react-intersection-observer`. But we've deprecated lodash in Calypso because, well, why have 5 different versions of mapping when `.map` and `=>` are already built into the language? Sure it's not as terse as using lodash helpers, but just "using the platform" keeps the API surface area low. Devs are going to have to learn some pattern, so make them learn the native one i.e. you will have to learn either `_.isEmpty( arr )` or `! Boolean( arr.length )` so you might as well learn the native one since it's teaching you how the language works. [The `left-pad` debacle](https://www.theregister.com/2016/03/23/npm_left_pad_chaos/) is like the extreme version of this I guess.

Now I admit in this case there's an argument for having something that wraps the intersection observer API since it's very imperative and React is going to need something hook based.

---

Before typing this out I was starting to think that maybe `react-intersection-observer` isn't such a big deal and we should just keep it. But I think I've talked myself back into my original position that it's better to err on the side of fewer dependencies.

p-jackson commented:
(p.s. I intentionally haven't reviewed the code in this PR because I think we should decide, independent of implementation, whether it's the right call)

zaguiini commented on client/lib/use-in-view/index.ts:
```diff
@@ -0,0 +1,46 @@
+
+			oneTimeCallback();
+
+			viewedRef.current = true;
```
Does it make sense to not connect the intersection observer if the callback has already been fired? I think it'll save some cycles as there will be no observer installed at all.

mpkelly replied:
I think this would only be beneficial if the component unmounts, so I didn't make a change. There are already a few optimisations in place.

zaguiini reviewed (commented):
Thanks, @mpkelly @p-jackson.

Sure, it makes sense given the arguments. Let's go with our in-house solution, then. The only thing I'd ask is to add tests so there are no surprises when modifying the code.

mpkelly commented:
@zaguiini, ok, I will look at adding some tests for the hook.

mpkelly committed:
Add some tests to verify that:
- the callback only fires when in view
- it never fires more than once regardless of total times in view

mpkelly commented:
@zaguiini, I have added a simple test suite that can verify that the callback only ever fires once when the element is in view. Please give another review.

danielbachhuber commented on client/lib/use-in-view/index.ts:
```diff
@@ -0,0 +1,46 @@
+
+		const handler = ( entries: IntersectionObserverEntry[] ) => {
+			// Only fire once per page load
+			if ( viewedRef.current ) {
```
What if someone wants this to fire every time the element is in view?

mpkelly replied:
It would require an update but just a small one. Do we have that use case currently?

danielbachhuber replied:
Not sure, haven't looked through the codebase.

Personally, I realize I'm strongly against reinventing the wheel in this case, so I'm going to refrain from reviewing this pull request further.

zaguiini reviewed (approved):
This works for me 👍 

I think the main benefit of having this in our codebase is that it'll be quicker to work on changes if necessary. I was initially hesitant but turns out it's not a big deal to have this in-house solution. Let's see if it works out -- we can always go back if the results are not great.
